table/block_based/block_based_table_builder.cc

```diff
@@ -1101,26 +1333,56 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
                 r->compression_type == kZSTDNotFinalCompression));
 
   for (size_t i = 0; ok() && i < r->data_block_and_keys_buffers.size(); ++i) {
-    const auto& data_block = r->data_block_and_keys_buffers[i].first;
+    auto& data_block = r->data_block_and_keys_buffers[i].first;
     auto& keys = r->data_block_and_keys_buffers[i].second;
     assert(!data_block.empty());
     assert(!keys.empty());
 
-    for (const auto& key : keys) {
-      if (r->filter_builder != nullptr) {
-        size_t ts_sz =
-            r->internal_comparator.user_comparator()->timestamp_size();
-        r->filter_builder->Add(ExtractUserKeyAndStripTimestamp(key, ts_sz));
+    if (r->compression_opts.parallel_threads > 1) {
+      ParallelCompressionRep::BlockRep block_rep;
+      r->pc_rep->block_rep_pool.pop(block_rep);
+
+      std::swap(*(block_rep.data_ptr), data_block);
+      block_rep.contents = *(block_rep.data_ptr);
+
+      block_rep.compression_type = r->compression_type;
+
+      block_rep.keys_ptr->SwapAssign(keys);
+
+      if (i + 1 < r->data_block_and_keys_buffers.size()) {
+        block_rep.first_key_in_next_block_ptr->assign(
+            r->data_block_and_keys_buffers[i + 1].second.front());
+      } else {
+        if (r->first_key_in_next_block_ptr == nullptr) {
+          delete block_rep.first_key_in_next_block_ptr;
+          block_rep.first_key_in_next_block_ptr = nullptr;
+        } else {
+          block_rep.first_key_in_next_block_ptr->assign(
+              r->first_key_in_next_block_ptr->data(),
+              r->first_key_in_next_block_ptr->size());
+        }
+      }
+
+      r->pc_rep->write_queue.push(std::move(block_rep.slot_ptr));
```

Comment:
I'm a little bit confused here. We moved `block_rep.slot_ptr`, but we still seem to continue using and cleaning up this pointer. It seems to be contradicting. Do we need `std::move()` here?

Reply:
WorkQueue's push method only accepts r-value by design. We have to use std::move to wrap the pointer, but the data referenced by pointer is not moved, only pointer value, i.e. a scalar, is "moved". Actually, the pointer value is just copied into the queue.

slot_ptr is now made a unique_ptr, and unique_ptr::get() returns a r-value itself, so std::move is not necessary. But std::move for block_rep's is still needed, because it's a l-value.